### ISSUE TITLE
pip install simpeg in windows

### ISSUE THREAD

**Issue by sgkang** [simpeg/simpeg#432]
I was installing simpeg using pip install through git repo 

so, I did
`pip install git+git://github.com/simpeg/simpeg.git`

and the below error is that I have got. I guess that is because of C ++ compiler, and possibly 
just because of the path. Any ideas?

C:\Users\administrator\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:S:\EOSC350\anaconda\libs /LIBPATH:S:\EOSC350\anaconda\PCbuild\amd64 /LIBPATH:S:\EOSC350\anaconda\PC\VS9.0\amd64 /EXPORT:initSimPEG/Utils/interputils_cython build\temp.win-amd64-2.7\Release\SimPEG/Utils/interputils_cython.obj /OUT:S:\EOSC350\pythonprograms\simpeg\SimPEG\Utils\interputils_cython.pyd /IMPLIB:build\temp.win-amd64-2.7\Release\SimPEG/Utils\interputils_cython.lib /MANIFESTFILE:build\temp.win-amd64-2.7\Release\SimPEG/Utils\interputils_cython.pyd.manifest
LINK : error LNK2001: unresolved external symbol initSimPEG/Utils/interputils_cython
build\temp.win-amd64-2.7\Release\SimPEG/Utils\interputils_cython.lib : fatal error LNK1120: 1 unresolved externals
error: command 'C:\Users\administrator\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\link.exe' failed with exit status 1120

**Comment by lheagy** [comment]
Brandon encountered a linker error with windows at BIRS. Here are his notes: 

**Notes on SimPEG Installation** 
RE: Microsoft Visual C++ Compiler Issues
B. Karchewski
20 August 2016

System:
Windows 7 Enterprise x64 PC
Intel Core i7-5600 2.60GHz CPU
msys MinGW for Unix-like commands
Anaconda python 2.7 build

Steps:

1. Clone SimPEG from git repo, git clone https://github.com/simpeg/simpeg
2. Install SimPEG using setup.py, python setup.py install
	-Error encountered: Missing linker, needs MSC v.1500 (Microsoft Visual C++ 2008) Runtime Library
	-Available: https://www.microsoft.com/en-ca/download/details.aspx?id=29
	-Apparently the issue was related to requiring the same set of MSC runtime libraries
		that were used to compile the python distribution that is installed
	-Why python would build using any MSC rather than an open-source C compiler such as gcc or g++
		is beyond me, but apparently there are dependencies on the MSC
3. After installing the redistributable run-time libraries, repeated Step 2 and installation seemed
	to work

**Comment by jcapriot** [comment]
I had an issue installing it on my windows machine using python 3.5 while having only the compatible msvcc 2015 libraries and compilers on my computer, which was odd because I had compiled other things for python just fine. I think it came down to a improper linking command, which i didn't have time to debug. I had rewrote the setup.py script to use numpy's distutils package, as well as the same format used in scipy's source code and it worked just fine without me having to change any of my compilers.

**Comment by sgkang** [comment]
Thanks guys for comments!
Few questions:

@lheagy what is "the redistributable run-time libraries"?

@jcapriot "setup.py" is for simpeg? then is there a chance that we can make sure that is working when we are pip installing?

**Comment by maguman** [comment]
Hello? I encountered the same error during setup.
Haven't you found any solution yet?

**Comment by maguman** [comment]
I solved my problem described above.
I think the **directory traversal character** is key to my problem.

**Comment by dhanssen** [comment]
@maguman , which steps did you take to solve this problem?
I receive the same error, even after I followed the different steps posted by @lheagy 

LINK : error LNK2001: unresolved external symbol initSimPEG/Utils/interputils_cython
    build\temp.win-amd64-2.7\Release\SimPEG/Utils\interputils_cython.lib : fatal error LNK1120: 1 unresolved externals
    error: command 'C:\\Users\\daan\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\amd64\\link.exe' failed with exit status 1120

**Comment by maguman** [comment]
@dhanssen Give me some time to examine my steps and summarize my case.

**Comment by dhanssen** [comment]
@maguman Too bad! But, thank you anyway!
Let's hope I can figure it out as well.
Or I might just go for that Linux dual-boot, we'll see.

**Comment by maguman** [comment]
@dhanssen Try to follow the steps posted by @lheagy but at this time, edit the follow lines 69-72 on 'setup.py' before taking Step 2.

Before:
cython_files = [
                    "SimPEG/Utils/interputils_cython",
                    "SimPEG//Mesh//TreeUtils"
               ] 

After:
cython_files = [
                    "SimPEG//Utils//interputils_cython",
                    "SimPEG//Mesh//TreeUtils"
               ]

**Comment by maguman** [comment]
@dhanssen 

**Notes on SimPEG Installation**
Windows 7 Issues
Y. Chung
7 December 2016

System:
Windows 7 Enterprise x64 PC

Symtoms:
During the installment either by "pip install SimPEG" or "python setup.py install" we encountered following error,

LINK : error LNK2001: unresolved external symbol initSimPEG/Utils/interputils_cython
build\temp.win-amd64-2.7\Release\SimPEG/Utils\interputils_cython.lib : fatal error LNK1120: 1 unresolved externals
error: command 'C:\Users\daan\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\link.exe' failed with exit status 1120

Remedy:

- Clone SimPEG from git repo,  git clone https://github.com/simpeg/simpeg
- Open jupyter notebook, then edit setup.py at lines 69-72  as shown below.
- Install SimPEG using setup.py, python setup.py install.
- If you get errors, ignore the errors and just repeat Step 3 several times until it works.

```
Before:
cython_files = [
"SimPEG/Utils/interputils_cython",
"SimPEG/Mesh/TreeUtils"
]

After:
cython_files = [
"SimPEG//Utils//interputils_cython",
"SimPEG//Mesh//TreeUtils"
]
```

**Comment by maguman** [comment]
@dhanssen 

I realized that I had modified one more file, that is "dependencies.py". lines 72-99.
I replaced "slash" with "back-slash".

```
Before:
def extended_iglob(pattern):
    if '{' in pattern:
        m = re.match('(.*){([^}]+)}(.*)', pattern)
        if m:
            before, switch, after = m.groups()
            for case in switch.split(','):
                for path in extended_iglob(before + case + after):
                    yield path
            return
    if '**/' in pattern:
        seen = set()
        first, rest = pattern.split('**/', 1)
        if first:
            first = iglob(first+'/')
        else:
            first = ['']
        for root in first:
            for path in extended_iglob(join_path(root, rest)):
                if path not in seen:
                    seen.add(path)
                    yield path
            for path in extended_iglob(join_path(root, '*', '**/' + rest)):
                if path not in seen:
                    seen.add(path)
                    yield path
    else:
        for path in iglob(pattern):
            yield path

After:
def extended_iglob(pattern):
    if '{' in pattern:
        m = re.match('(.*){([^}]+)}(.*)', pattern)
        if m:
            before, switch, after = m.groups()
            for case in switch.split(','):
                for path in extended_iglob(before + case + after):
                    yield path
            return
    if '**\' in pattern:
        seen = set()
        first, rest = pattern.split('**\', 1)
        if first:
            first = iglob(first+'\')
        else:
            first = ['']
        for root in first:
            for path in extended_iglob(join_path(root, rest)):
                if path not in seen:
                    seen.add(path)
                    yield path
            for path in extended_iglob(join_path(root, '*', '**\' + rest)):
                if path not in seen:
                    seen.add(path)
                    yield path
    else:
        for path in iglob(pattern):
            yield path
```

Since I am new to python and cython, my remedy is not standard. I just focused on parsing parts of the source codes. As is well known, we use '/' on linux while we use '\' on windows.

**Comment by lheagy** [comment]
Sorry for the delay in response, yes it does look like an issue with file-separations in the setup.py. 

We should be able to resolve this by replacing
```
cython_files = [
                    "SimPEG/Utils/interputils_cython",
                    "SimPEG/Mesh/TreeUtils"
               ]
```
with 
```python
cython_files = [
                    "SimPEG{sep}Utils{sep}interputils_cython".format(sep=os.path.sep),
                    "SimPEG{sep}Mesh{sep}TreeUtils".format(sep=os.path.sep)
               ]
```
@rowanc1: can con correct me if you think this is off-base?

**Comment by ahartikainen** [comment]
Fix path issues with `os.path.join`. Works the correct way on linux, mac and windows.

    `os.path.join('SimPEG', 'Utils', 'interputils_cython')`

SimPEG and general folders can even be variables.

**Comment by lheagy** [comment]
I believe this should have closed with #542, @rowanc1 : can you confirm?

**Comment by rowanc1** [comment]
We have removed these files and file pointers when we pulled out the discretize package last month. @jcapriot also spent some time working on the `setup.py` file to be more in line with the `numpy` package, which has multiple `setup.py`'s in each of the sub-packages.

We are now testing the discretize package on windows on every commit, which should help remedy some of the more obvious bugs (like `path.sep/join`).

We still have a few outstanding issues with the deployment in that it is not going from python 3.0 (#560) and that the wheels are not being built/deployed for windows users (simpeg/discretize#26).

I am going to close this now, however, please let us know if anyone has any other problems on Windows/Pypi!
